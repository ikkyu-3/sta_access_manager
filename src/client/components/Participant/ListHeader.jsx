// @flow
import * as React from "react";
import styles from "./style.css";

function ListHeader(): React.Node {
  const { header, columnName, columnPurpose, columnStatus, border } = styles;
  return (
    <div className={header}>
      <div className={`${columnName} ${border}`}>名前</div>
      <div className={`${columnPurpose} ${border}`}>目的</div>
      <div className={columnStatus}>入退室</div>
    </div>
  );
}
export default ListHeader;
