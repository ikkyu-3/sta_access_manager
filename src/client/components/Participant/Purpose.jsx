// @flow
import type { ParticipantPurposeProps } from "@types";

import * as React from "react";
import { MEET_UP, STUDY, WORK, CIRCLE } from "@constants";
import styles from "./style.css";

function Purpose(props: ParticipantPurposeProps): React.Node {
  const { purpose } = props;
  const { columnPurpose, border, work, study, meetUp, circle } = styles;

  let background: string = "";
  let value: string = "";

  if (purpose === MEET_UP) {
    background = meetUp;
    value = "勉強会";
  }
  if (purpose === STUDY) {
    background = study;
    value = "自習";
  }
  if (purpose === WORK) {
    background = work;
    value = "仕事";
  }
  if (purpose === CIRCLE) {
    background = circle;
    value = "サークル";
  }

  return (
    <div className={`${columnPurpose} ${border} ${background}`}>{value}</div>
  );
}
export default Purpose;
