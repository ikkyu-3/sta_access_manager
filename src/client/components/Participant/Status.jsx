// @flow
import type { ParticipantStatusProps } from "@types";

import * as React from "react";
import styles from "./style.css";

function Status(props: ParticipantStatusProps): React.Node {
  const { isEntry } = props;
  const { columnStatus, entry, exit } = styles;

  const className = isEntry
    ? `${columnStatus} ${entry}`
    : `${columnStatus} ${exit}`;
  const value = isEntry ? "入" : "退";

  return <div className={className}>{value}</div>;
}
export default Status;
