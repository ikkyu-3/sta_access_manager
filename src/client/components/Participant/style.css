@charset "utf-8";
.flame {
  padding: 8px;
  background: white;
  border-radius: 8px;
  box-sizing: border-box;
}

.list {
  height: 100%;
  font-size: 24px;
}

.list .header {
  display: flex;
  height: 60px;
  font-weight: bold;
  border: 2px solid lightgray;
  box-sizing: border-box;
}

.list .header > div {
  display: flex;
  justify-content: center;
  align-items: center;
}

.columnName {
  flex-grow: 1;
}

.columnPurpose {
  width: 130px;
}

.columnStatus {
  width: 90px;
}

.border {
  border-right: 2px solid lightgray;
  box-sizing: border-box;
}

.list .body {
  height: calc(100% - 60px);
  border: 2px solid lightgray;
  border-top: transparent;
  box-sizing: border-box;
  overflow: scroll;
}

.list .body ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.list .body li {
  display: flex;
  height: 52px;
}

.list .body li.buttom {
  border-bottom: 2px solid lightgray;
}

.list .body li > div {
  display: flex;
  justify-content: center;
  align-items: center;
}

/* 仕事 */
.work {
  background-color: #bbfeff;
}

/* 自習 */
.study {
  background-color: #bcf3cb;
}

/* 勉強会 */
.meetUp {
  background-color: #f8f189;
}

/* サークル */
.circle {
  background-color: #f4d8f5;
}

.entry {
  background-color: #ffd400;
}

.exit {
  background-color: #ebebeb;
}
