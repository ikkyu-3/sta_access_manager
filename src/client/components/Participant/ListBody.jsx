// @flow
import type { ParticipantListBodyProps, User } from "@types";

import * as React from "react";
import Purpose from "./Purpose";
import Status from "./Status";
import styles from "./style.css";

function ListBody(props: ParticipantListBodyProps): React.Node {
  const { body, columnName, border, buttom } = styles;
  const { listData } = props;

  const { length } = listData;
  const listItems = listData.map((data: User, index: number): React.Node => {
    const { id, user } = data;
    const { name, purpose, isEntry } = user;
    const listButtom: string = index === length - 1 && index > 5 ? "" : buttom;
    const key = `list-${id}`;

    return (
      <li key={key} className={listButtom}>
        <div className={`${columnName} ${border}`}>{name}</div>
        <Purpose purpose={purpose} />
        <Status isEntry={isEntry} />
      </li>
    );
  });

  return (
    <div className={body}>
      <ul>{listItems}</ul>
    </div>
  );
}
export default ListBody;
