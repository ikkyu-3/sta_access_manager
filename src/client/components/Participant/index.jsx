// @flow
import type { ParticipantProps } from "@types";

import * as React from "react";
import ListHeader from "./ListHeader";
import ListBody from "./ListBody";
import styles from "./style.css";

class Participant extends React.Component<ParticipantProps> {
  componentDidMount() {
    this.props.refresh();
    this.props.selectedMenu();
    this.props.getParticipant();
  }

  render(): React.Node {
    const { flame, list } = styles;
    const { participant } = this.props;

    return (
      <div className={flame}>
        <div className={list}>
          <ListHeader />
          <ListBody listData={participant} />
        </div>
      </div>
    );
  }
}

export default Participant;
