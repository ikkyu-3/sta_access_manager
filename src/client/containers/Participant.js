// @flow
import type { State, Dispatch } from "@types";

import { connect } from "react-redux";
import { push } from "react-router-redux";

import {
  selectParticipant,
  requestStart,
  requestEnd,
  removeId,
  clearFirstName,
  clearLastName,
  clearMailAddress,
  hideInputValidationResult,
  setParticipantAction,
  removeParticipant
} from "@actions";
import Participant from "@components/Participant";
import {
  API_GET_USERS,
  ERROR_BAD_REQUEST_PATH,
  ERROR_INTERNAL_SERVER_ERROR_PAHT
} from "@constants";
import request from "@modules/request";
import createApiUrl from "@modules/createApiUrl";

const mapStateToProps = (state: State) => ({
  participant: state.participant
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  refresh: () => {
    dispatch(removeId());
    dispatch(clearFirstName());
    dispatch(clearLastName());
    dispatch(clearMailAddress());
    dispatch(hideInputValidationResult());
  },
  selectedMenu: () => {
    dispatch(selectParticipant());
  },
  getParticipant: async () => {
    const { method, path } = API_GET_USERS;
    const url: string = createApiUrl(path);
    try {
      dispatch(push(requestStart()));
      const res = await request({ url, method });
      dispatch(push(requestEnd()));
      // TODO: スキーマチェック
      if (res.data) {
        dispatch(setParticipantAction(res.data));
      } else {
        dispatch(removeParticipant());
        dispatch(push(ERROR_BAD_REQUEST_PATH));
      }
    } catch (e) {
      dispatch(push(requestEnd()));
      dispatch(removeParticipant());
      if (e.statusCode === 500) {
        dispatch(push(ERROR_INTERNAL_SERVER_ERROR_PAHT));
      } else {
        dispatch(push(ERROR_BAD_REQUEST_PATH));
      }
    }
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Participant);
