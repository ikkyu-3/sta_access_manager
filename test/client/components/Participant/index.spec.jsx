import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Participant from "@components/Participant";
import { shortParticipantData } from "../testData";

const fn = () => {};

describe("Participantのテスト", () => {
  describe("スナップショット", () => {
    it("正しいレンダリング", () => {
      const tree = renderer
        .create(
          <Participant
            refresh={fn}
            selectedMenu={fn}
            getParticipant={fn}
            participant={shortParticipantData}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("コンポーネントのテスト", () => {
    const refreshMock = jest.fn();
    const selectedMenuMock = jest.fn();
    const getParticipantMock = jest.fn();
    let participant;

    beforeAll(() => {
      participant = shallow(
        <Participant
          refresh={refreshMock}
          selectedMenu={selectedMenuMock}
          getParticipant={getParticipantMock}
          participant={shortParticipantData}
        />
      );
    });

    it("componentDidMountは、refresh, selectedMenu, getParticipantを実行する", () => {
      expect(refreshMock).toBeCalled();
      expect(selectedMenuMock).toBeCalled();
      expect(getParticipantMock).toBeCalled();
    });

    it("ListBodyコンポーネントのlistDataに、dataで指定した値が渡されている", () => {
      expect(participant.find("ListBody").prop("listData")).toBe(
        shortParticipantData
      );
    });
  });
});
