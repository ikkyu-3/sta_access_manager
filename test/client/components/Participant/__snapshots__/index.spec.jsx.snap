// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Participantのテスト スナップショット 正しいレンダリング 1`] = `
<div
  className="flame"
>
  <div
    className="list"
  >
    <div
      className="header"
    >
      <div
        className="columnName border"
      >
        名前
      </div>
      <div
        className="columnPurpose border"
      >
        目的
      </div>
      <div
        className="columnStatus"
      >
        入退室
      </div>
    </div>
    <div
      className="body"
    >
      <ul>
        <li
          className="buttom"
        >
          <div
            className="columnName border"
          >
            name1
          </div>
          <div
            className="columnPurpose border "
          >
            
          </div>
          <div
            className="columnStatus entry"
          >
            入
          </div>
        </li>
        <li
          className="buttom"
        >
          <div
            className="columnName border"
          >
            name2
          </div>
          <div
            className="columnPurpose border "
          >
            
          </div>
          <div
            className="columnStatus entry"
          >
            入
          </div>
        </li>
        <li
          className="buttom"
        >
          <div
            className="columnName border"
          >
            name3
          </div>
          <div
            className="columnPurpose border "
          >
            
          </div>
          <div
            className="columnStatus entry"
          >
            入
          </div>
        </li>
        <li
          className="buttom"
        >
          <div
            className="columnName border"
          >
            name4
          </div>
          <div
            className="columnPurpose border "
          >
            
          </div>
          <div
            className="columnStatus entry"
          >
            入
          </div>
        </li>
        <li
          className="buttom"
        >
          <div
            className="columnName border"
          >
            name5
          </div>
          <div
            className="columnPurpose border "
          >
            
          </div>
          <div
            className="columnStatus exit"
          >
            退
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
